node/src/main.rs

```diff
@@ -169,6 +179,40 @@ fn async_main() -> impl Future<Item = (), Error = ()> + Send + 'static {
     // Forward schema events from the subgraph provider to the schema provider
     tokio::spawn(forward(&mut subgraph_provider, &schema_provider).unwrap());
 
+    // Start admin JSON-RPC server.
+    let json_rpc_server =
+        JsonRpcServer::serve(json_rpc_port, Arc::new(subgraph_provider), logger.clone())
+            .expect("Failed to start JSON-RFC server");
+
+    // Let the server run forever.
+    std::mem::forget(json_rpc_server);
+
+    // Add the CLI subgraph with a REST request to the admin server.
+    if let Some(subgraph) = subgraph {
+        let (name, hash) = if subgraph.contains(':') {
+            let mut split = subgraph.split(':');
+            (split.next().unwrap(), split.next().unwrap())
+        } else {
+            ("cli", subgraph)
+        };
+
+        let raw_response = Client::new()
+            .post(&format!(
+                "http://localhost:{}/subgraph_add/{}/{}",
```

Comment:
Hmm, is there a solution here that wouldn't require formatting a URL? I don't think it matters _too_ much, but the injection possibilities here make me nervous. E.g. if the command line arg is "../whatever-command-i-want:whatever", then that does surprising things

Reply:
I second that.

Reply:
While it's possible refactor this into some sort of direct method call, for maintainability and dogfooding it's nice that this goes through the same code path as any other `subgraph_add` request. Also we're likely to deprecate adding subgraphs through the CLI before shipping.

Do you think it would be sufficient to parse the URL and check that:
- the scheme is http
- The host is `localhost`
- The port is `json_rpc_port`
- It has three path segments, respectively `subgraph_add`, `name` and `hash`

Reply:
I guess this is less about checking the above properties and more about using something similar to `path.join("foo", "bar", "baz")` instead of building the URL from a format string. which would make sure that if a value like `"foo/bar"` was passed in, the slash would be escaped / URL encoded.

Reply:
Still not enough, try name = ".." and hash = "any-command" :)

What about using the JSON-RPC style instead? It seems more robust

Reply:
In Rust, `Url::parse` (for the fixed `"http://localhost/"` part), `set_port` and `path_segments_mut()` (https://docs.rs/url/1.7.1/url/struct.Url.html#method.path_segments_mut) may provide what we want.

Reply:
@timmclean I don't follow what attack you are envisioning there, do you mean trying to run shell commands from an http request?

@Jannis that seems like a good way to make sure we're building the URL we think we're building. @timmclean do you think that would be sufficient?

Reply:
It looks like `PathSegmentsMut::extend` ignores `"."` and `".."`, so that might work!

No specific attack possible yet. This is just about a hypothetical scenario where we extend the JSON-RPC interface with more endpoints, and this code gets used with input parameters that aren't trusted. In that case, a special input could convince the code to do an HTTP request that does something other than `subgraph_add`.

Is there any reason we can't use the normal JSON-RPC style with a POST request body like `{"jsonrpc":"2.0","method": ...}`? I find path and URL manipulations are just so hard to get right, even with the help of a good library

Reply:
Thanks, I'll follow your advice and try doing a full json-rpc request. For my knowledge, what's the issue with `..` in urls?

Reply:
Just a little trick to mess up the path processing. So, if you supply `..` and `do_self_destruct` as parameters, then that will create the URL `http://localhost:8020/subgraph_add/../do_self_destruct`. When the server processes that request, most servers will take `/subgraph_add/../do_self_destruct` and simplify it to `/do_self_destruct`, because `..` means "parent directory". So then it runs `do_self_destruct` instead of `subgraph_add` :)

Reply:
Ah I see! Scary.